Feature/customcolumns

cryptomental opened the pull request:
Example usage:

a) only identifier and tags
```
$ piston list --limit 1 --columns identifier _tags                                                          
+----------------------------------------------------------+-------------------------------------------+
| identifier                                               | _tags                                     |
+----------------------------------------------------------+-------------------------------------------+
| @steve-walschot/are-we-only-here-to-feed-the-narcissists | ['steemit', 'dollarvigilante', 'worries'] |
+----------------------------------------------------------+-------------------------------------------+
```

b) only identifier, title and tags
```
$ piston list --limit 1 --columns identifier title _tags
+----------------------------------------------------------+-------------------------------------------+-------------------------------------------+
| identifier                                               | title                                     | _tags                                     |
+----------------------------------------------------------+-------------------------------------------+-------------------------------------------+
| @steve-walschot/are-we-only-here-to-feed-the-narcissists | Are we only here to feed the narcissists? | ['steemit', 'dollarvigilante', 'worries'] |
+----------------------------------------------------------+-------------------------------------------+-------------------------------------------+
```

c) only identifier and title

```
$ piston list --limit 3 --columns identifier title
+------------------------------------------------------------+--------------------------------------------------------------+
| identifier                                                 | title                                                        |
+------------------------------------------------------------+--------------------------------------------------------------+
| @steve-walschot/are-we-only-here-to-feed-the-narcissists   | Are we only here to feed the narcissists?                    |
| @anyx/cheetah-bot-the-fight-against-spam-and-plagiarism-   | Cheetah bot: The Fight Against Spam and Plagiarism Continues |
|  continues                                                 |                                                              |
| @liberosist/mind-your-votes-an-investigation-and-guide-to- | Mind Your Votes! An investigation and guide to maximizing    |
|  maximizing-your-curation-rewards                          |  your Curation Rewards.                                      |
+------------------------------------------------------------+--------------------------------------------------------------+
```

cryptomental commented:
Hi I need to rebase, please do not merge yet!

cryptomental commented:
@xeroc the pull request was rebased and should be fine now. I had different 'git commiter' and 'git author' set in my commits due to my local setup.

xeroc commented on piston/ui.py:
```diff
@@ -104,37 +108,114 @@ def wrapText(t):
-            d["children"],
-            # d["net_rshares"],
-            d["pending_payout_value"],
+def __get_discussion_identifier(discussion):
```
There is a `constructIdentifier` in `utils.py`

cryptomental replied:
ok thanks I will use it then.

cryptomental replied:
Done. I am using constructIdentifier now.

xeroc commented on piston/ui.py:
```diff
@@ -104,37 +108,114 @@ def wrapText(t):
+            ]
+            t.add_row(row)
+    else:
+        allowed_columns = {"_tags", "active", "cashout_time", "category",
```
Why do you restrict users to certain attributes?

cryptomental replied:
Good question, would you propose to extend the attributes list? If so, where can I find the complete list that should be available? I printed out discussion.__dict__  but the list is huge and I was not sure how much to expose.

xeroc replied:
Well, it's easier to maintain if you just don't restrict the user .. if they want a huge output that cripples their screen, so be it .. to ensure an attrribute exists before using it, you can use

    hasattr(post, "attribute")

or simply

    getattr(post, "attribute", None)

cryptomental replied:
Thank you, I will implement that.

cryptomental commented:
@xeroc I updated the pull request to use constructIdentifier and rebased my changes. The question remains about the column list that should be available.

xeroc commented:
Have you every 'squashed' commits into a single commit? If you, please do so, if not, let me know I can guide you or do it for you .. Sicne this is basically a single feature, it could be squashed into a single commit aswell ..

cryptomental commented:
@xeroc I now squashed all commits into a single commit, please have a look if the commit message format is ok for you or I should change it.

I think I solved the attrs with quite an elegant solution by comparing against vars() output.

Now the output is:

```
$ piston list --category introduceyourself --sort cashout --limit 100 --columns identifier _tags created total_payout_reward aaa bbb
Traceback (most recent call last):
  File "/usr/local/bin/piston", line 11, in <module>
    sys.exit(main())
  File "/home/piston/.local/lib/python3.5/site-packages/steem_piston-0.3-py3.5.egg/piston/__main__.py", line 931, in main
    args.columns
  File "/home/piston/.local/lib/python3.5/site-packages/steem_piston-0.3-py3.5.egg/piston/ui.py", line 174, in list_posts
    (sorted(available_attrs), wrong_columns))
piston.ui.UIError: Please use allowed column names only: ['_json_metadata', '_patch', '_tags', 'abs_rshares', 'active', 'active_parsed', 'active_votes', 'allow_curation_rewards', 'allow_replies', 'allow_votes', 'author', 'author_rewards', 'body', 'cashout_time', 'cashout_time_parsed', 'category', 'children', 'children_abs_rshares', 'children_rshares2', 'created', 'created_parsed', 'curator_payout_value', 'depth', 'id', 'identifier', 'json_metadata', 'last_payout', 'last_payout_parsed', 'last_update', 'last_update_parsed', 'max_accepted_payout', 'max_cashout_time', 'max_cashout_time_parsed', 'net_rshares', 'net_votes', 'openingPostIdentifier', 'parent_author', 'parent_permlink', 'pending_payout_value', 'percent_steem_dollars', 'permlink', 'replies', 'reward_weight', 'root_comment', 'root_title', 'steem', 'title', 'total_payout_reward', 'total_payout_value', 'total_pending_payout_value', 'total_vote_weight', 'url', 'vote_rshares']. Error caused by {'aaa', 'bbb'}.
```

xeroc commented on piston/ui.py:
```diff
@@ -104,37 +108,95 @@ def wrapText(t):
+            t.add_row(row)
+    else:
+        available_attrs = set(vars(discussions[0]))
+        if not set(custom_columns).issubset(available_attrs):
```
nice ..

cryptomental commented:
I noticed one corner case when the discussion list is empty, will fix this soon.

```
$ piston list --category introduceyourself --sort payout --limit 100 --columns identifier _tags pending_payout_value
Traceback (most recent call last):
  File "/usr/local/bin/piston", line 11, in <module>
    sys.exit(main())
  File "/home/piston/.local/lib/python3.5/site-packages/steem_piston-0.3-py3.5.egg/piston/__main__.py", line 931, in main
    args.columns
  File "/home/piston/.local/lib/python3.5/site-packages/steem_piston-0.3-py3.5.egg/piston/ui.py", line 169, in list_posts
    available_attrs = set(vars(discussions[0]))
IndexError: list index out of range
```

cryptomental commented:
Done, I left the commit message as is for you to review.

xeroc commented:
Good job!

cryptomental commented:
Thanks!
